61 window expansion buowset1.1

kgarwoodsdzwa opened the pull request:
added a function to offer the option for window expansion when generating the detection segments, can be even or random. the duration can be changed in the /whoot package function definition. used to generate buowset 1.1

kgarwoodsdzwa committed:
added function to do window expansion to the audio segments

to the function that creates the vocalization segment i just added
a new function that adds equal time to the beginning and end to
get the segment to 3s surrounding the actual duration of the
vocalization

kgarwoodsdzwa committed:
add function needed to init instead because it might be useful

the window expansion function could be useful elsewhere so in an
attempt to packagize im putting it here and it could be useful when
preparing other datasets later

kgarwoodsdzwa committed:
change the create dataset scripts to use the new function in init

the window expansion function was added to the package itself, so
the scripts that needed it now call it from there to do the
window expansion option when one goes to create the buowset dataset.
one can choose from window expansion where the beginning and end are
expanded equally to hit 3s and also random expansion where it still
hits 3s but randomly expands the beginning and end

kgarwoodsdzwa committed:
fixing some bugs found after testing

had to change the variable random to randomize so it wouldnt
conflict with the random package invocation, and had to allow for
no window expansion in case the audio segment isn't long
enough to randomly window expand

kgarwoodsdzwa committed:
fixed usage line

i have to run the script as a module now if i want it to use the
window expansion function in whoot __init__

Sean1572 commented on whoot/audio_utils.py:
```diff
@@ -0,0 +1,47 @@
+        try: 
+            segment = audio[new_start:new_end]
+        except Exception as e:
+            segment = audio[start_time:end_time]
```
If the file clip is long enough, instead of taking the original time, take the first n seconds of a clip / last n seconds of a clip, depending on whether end_time > clip_length or if start_time < 0

Then handle the case where clip_length < desired window length

kgarwoodsdzwa replied:
i reread this many times and i think i fixed it? let me know

Sean1572 reviewed (changes requested):
Must have approved by mistake this morning

kgarwoodsdzwa committed:
linting

kgarwoodsdzwa committed:
linting

kgarwoodsdzwa committed:
handle extra cases

handle case where the audio file is shorter than the desired clip
length, and the case where the duration of the vocalization is
longer than the desired clip length

kgarwoodsdzwa committed:
add another case handler

handle the case where the window expansion wants to expand past the
end time of the clip, wherein it will just end the clip at the end
of the audiofile and backtrack out the start time

Sean1572 commented on whoot/audio_utils.py:
```diff
@@ -0,0 +1,58 @@
+    if randomize:
+        diff = length-duration
+        offset = random.uniform(0, diff)
+        new_start = start_time-offset
```
If start_time < offset, then new_start < 0.  That could produce some weird behavior when indexing the audio file.

Sean1572 replied:
Basically, you kinda have to do the same thing you do to the case where new_end > clip_length, except the other way around (when new_start < 0)

kgarwoodsdzwa committed:
add handling of start before 0

kgarwoodsdzwa committed:
adding return value for the relative start time of the strong label

kgarwoodsdzwa committed:
add a little linting

kgarwoodsdzwa committed:
adding the rel_start in metadata if window expansion is involved

kgarwoodsdzwa committed:
fix row mismatch when adding the relative seg start

kgarwoodsdzwa commented:
make the segment duration column fill, and fill in the segment rel start as 0 if the clip was already 3s.

kgarwoodsdzwa commented:
scratch that- just filling in the no buow segment rel start ms with 0

kgarwoodsdzwa committed:
change imports and name of one column for clarity

Sean1572 reviewed (approved):
LGTM
